// appium-xcuitest-driver/build/lib/device-log/safari-console-log.js
'use strict';

var _createClass = require('babel-runtime/helpers/create-class')['default'];

var _classCallCheck = require('babel-runtime/helpers/class-call-check')['default'];

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _getIterator = require('babel-runtime/core-js/get-iterator')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _appiumSupport = require('appium-support');

var MAX_LOG_ENTRIES_COUNT = 10000;

var log = _appiumSupport.logger.getLogger('SafariConsole');
log.warning = log.warn; // js console has `warning` level, so map to `warn`

var SafariConsoleLog = (function () {
  function SafariConsoleLog(showLogs) {
    _classCallCheck(this, SafariConsoleLog);

    this.showLogs = showLogs;
    this.logs = [];
    this.logIdxSinceLastRequest = 0;

    this.isCapturing = false;
  }

  _createClass(SafariConsoleLog, [{
    key: 'startCapture',
    value: function startCapture() {
      return _regeneratorRuntime.async(function startCapture$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            this.isCapturing = true;

          case 1:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'stopCapture',
    value: function stopCapture() {
      return _regeneratorRuntime.async(function stopCapture$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            this.isCapturing = false;

          case 1:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'addLogLine',
    value: function addLogLine(out) {
      if (this.isCapturing) {
        this.logs = this.logs || [];
        while (this.logs.length >= MAX_LOG_ENTRIES_COUNT) {
          this.logs.shift();
          if (this.logIdxSinceLastRequest > 0) {
            this.logIdxSinceLastRequest--;
          }
        }
        this.logs.push(out);
      }

      // format output like
      //     SafariConsole [WARNING][http://appium.io 2:13] Log something to warn
      if (this.showLogs) {
        var level = 'debug';
        if (out.level === 'warning' || out.level === 'error') {
          level = out.level;
        }
        var _iteratorNormalCompletion = true;
        var _didIteratorError = false;
        var _iteratorError = undefined;

        try {
          for (var _iterator = _getIterator(out.text.split('\n')), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
            var line = _step.value;

            // url is optional, so get formatting here
            var url = out.url ? out.url + ' ' : '';
            log[level]('[' + level.toUpperCase() + '][' + url + out.line + ':' + out.column + '] ' + line);
          }
        } catch (err) {
          _didIteratorError = true;
          _iteratorError = err;
        } finally {
          try {
            if (!_iteratorNormalCompletion && _iterator['return']) {
              _iterator['return']();
            }
          } finally {
            if (_didIteratorError) {
              throw _iteratorError;
            }
          }
        }
      }
    }
  }, {
    key: 'getLogs',
    value: function getLogs() {
      var result;
      return _regeneratorRuntime.async(function getLogs$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            if (!(this.logs.length && this.logIdxSinceLastRequest < this.logs.length)) {
              context$2$0.next = 5;
              break;
            }

            result = _lodash2['default'].clone(this.logs);

            if (this.logIdxSinceLastRequest > 0) {
              result = result.slice(this.logIdxSinceLastRequest);
            }
            this.logIdxSinceLastRequest = this.logs.length;
            return context$2$0.abrupt('return', result);

          case 5:
            return context$2$0.abrupt('return', []);

          case 6:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'getAllLogs',
    value: function getAllLogs() {
      return _regeneratorRuntime.async(function getAllLogs$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            return context$2$0.abrupt('return', _lodash2['default'].clone(this.logs));

          case 1:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }]);

  return SafariConsoleLog;
})();

exports.SafariConsoleLog = SafariConsoleLog;
exports['default'] = SafariConsoleLog;
